Bridged usdc custom gateway

gvladika opened the pull request:
Custom USDC gateway (for parent and child chain), complying with Bridged USDC Standard. These contracts can be used by new Orbit chains which want to provide USDC bridging solution and keep the possibility to upgrade to native USDC at some point later. This solution will NOT be used in existing Arbitrum chains.

Reference to the Circle's Bridged USDC: https://github.com/circlefin/stablecoin-evm/blob/master/doc/bridged_USDC_standard.md

Expected flow of actions on a fresh Orbit chain could look like this:

- chain owner deploys upgradeable USDC token on parent chain, or uses canonical USDC if there is one
- chain owner deploys upgradeable USDC token on child chain, to be used ad bridged USDC, and inits the contract
- chain owner deploys upgradeable USDC custom gateway on parent and child chain
    - chain owner inits those gateways
- chain owner registers USDC gateway in router for USDC token
- chain owner assigns minter role to the child chain gateway along with max mint allowance
- now USDC can be bridged in both directions
- once transition to native USDC is agreed on, chain owner pauses deposits on parent chain
- chain owner then pauses withdrawals on child chain
- chain owner removes the minter role from the child chain gateway
    - there should be no in-flight deposits when minter role is revoked. If there are any, they should be executed (can be done by anyone by claiming the failed retryable ticket which does the USDC depositing).
    - if minter role is revoked before in-flight deposits are claimed, those funds can’t be minted. One option is to leave the gateway’s minter role, but decrease the allowance to match the total amount of unclaimed deposits
- chain owner sets Circle as burner on the parent chain gateway using `setBurner(address)`  function
- chain owner reads the total supply of USDC on child chains. Then, he invokes `setBurnAmount(uint256)` on the parent child gateway where the amount matches the total supply
    - in case there are unclaimed deposits, their total amount should be added to the supply as those tokens shall eventually be minted by child chain gateway
- USDC masterMinter gives minter role with 0 allowance to parent chain gateway, so the burn can be executed
- on the child chain, gateway owner calls the `setUsdcOwnershipTransferrer(address)` to set the account (provided and controlled by Circle) which will be able to transfer the bridged USDC ownership and proxy admin
- if not already owned by gateway, bridged USDC owner transfers ownership to gateway, and proxy admin transfers admin rights to gateway
- Circle uses `usdcOwnershipTransferrer` account to trigger `transferUSDCRoles(address)` which will set caller as USDC proxy admin and will transfer USDC ownership to the provided address (according to standard [[requirements](https://github.com/circlefin/stablecoin-evm/blob/master/doc/bridged_USDC_standard.md#token-roles)](https://github.com/circlefin/stablecoin-evm/blob/master/doc/bridged_USDC_standard.md#token-roles))
- Circle calls `burnLockedUSDC()` on the L1 USDC gateway using `burner` account to burn the `burnAmount` of USDC
    - remaining USDC will be cleared off when remaining in-flight USDC withdrawals are executed, if any
- USDC token contract on the child chain is considered native now. Circle has ownership of the contract and can set minter, blacklister and other roles

PR containts USDC gateway implementation for parent chain and child chain, Foundry unit tests and Hardhat e2e test.

gzeoneth reviewed (changes requested):
to recap:
- need both eth and erc20 versions of this new gateway
- standardize the deployment of child chain bridged usdc
- have admin set in both parent and child chain
- no need to use retryables for admin pause

gvladika commented:
> to recap:
> 
> * need both eth and erc20 versions of this new gateway
> * standardize the deployment of child chain bridged usdc
> * have admin set in both parent and child chain
> * no need to use retryables for admin pause

Added referent child chain token implementation `BridgedUsdcCustomToken `. Made child gateway ownable and pausable by owner, removed cross-chain pause. Still need to add ero20 version

gzeoneth commented on contracts/tokenbridge/ethereum/gateway/L1FeeTokenUSDCCustomGateway.sol:
should be `L1OrbitUSDCCustomGateway.sol`?
probably can even skip `Custom`

gvladika replied:
Renamed:
L1USDCCustomGateway -> L1USDCGateway
L2USDCCustomGateway -> L2USDCGateway
L1FeeTokenUSDCCustomGateway -> L1OrbitUSDCGateway

gzeoneth commented on contracts/tokenbridge/ethereum/gateway/L1FeeTokenUSDCCustomGateway.sol:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    function _createRetryable(
+        address _inbox,
+        address _to,
+        address _refundTo,
+        address _user,
+        uint256 _totalFeeAmount,
+        uint256 _l2CallValue,
+        uint256 _maxSubmissionCost,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes memory _data
+    ) internal override returns (uint256) {
+        {
+            // Transfer native token amount needed to pay for retryable fees to the inbox.
+            // Fee tokens will be transferred from user who initiated the action - that's `_user` account in
+            // case call was routed by router, or msg.sender in case gateway's entrypoint was called directly.
+            address nativeFeeToken = IERC20Bridge(address(getBridge(_inbox))).nativeToken();
+            uint256 inboxNativeTokenBalance = IERC20(nativeFeeToken).balanceOf(_inbox);
+            if (inboxNativeTokenBalance < _totalFeeAmount) {
+                address transferFrom = isRouter(msg.sender) ? _user : msg.sender;
+                IERC20(nativeFeeToken).safeTransferFrom(
+                    transferFrom, _inbox, _totalFeeAmount - inboxNativeTokenBalance
+                );
+            }
+        }
+
+        return IERC20Inbox(_inbox).createRetryableTicket(
+            _to,
+            _l2CallValue,
+            _maxSubmissionCost,
+            _refundTo,
+            _user,
+            _maxGas,
+            _gasPriceBid,
+            _totalFeeAmount,
+            _data
+        );
+    }
```
this is now implemented multiple times in different Orbit version contracts, we should consider refactoring the code to use this as a library or base contract.

gzeoneth commented on test-foundry/L2USDCCustomGateway.t.sol:
```diff
@@ -0,0 +1,269 @@
+    event WithdrawalsPaused();
+}
+
+contract L2USDC is L2GatewayToken {
```
this should be a separate prod contract, not in test

gvladika replied:
Removed in separate PR - https://github.com/OffchainLabs/token-bridge-contracts/pull/88

gvladika commented on contracts/tokenbridge/test/BridgedUsdcCustomToken.sol:
```diff
@@ -0,0 +1,26 @@
+    function initialize(string memory name_, address l2Gateway_, address l1Counterpart_) public {
+        L2GatewayToken._initialize({
+            name_: name_,
+            symbol_: "USDC.e",
```
No more `BridgedUsdcCustomToken`, instead `FiatTokenArbitrumOrbitV2_2` will be used where name/symbol are provided at init

gzeoneth reviewed (commented):
lgtm, some comments
also new contracts can be added into storage test

gvladika commented on contracts/tokenbridge/arbitrum/gateway/L2USDCGateway.sol:
```diff
@@ -0,0 +1,208 @@
+        return true;
+    }
+
+    function _isValidTokenAddress(address _l1Address, address _expectedL2Address)
```
https://github.com/OffchainLabs/token-bridge-contracts/pull/87/commits/25672c7077deaa2eae1cbc94d95a0e48d9dd4362

gvladika commented on contracts/tokenbridge/ethereum/gateway/L1USDCGateway.sol:
```diff
@@ -0,0 +1,200 @@
+     * @notice Burns the USDC tokens escrowed in the gateway.
+     * @dev    Can be called by burner when deposits are paused and when
+     *         owner has set the burn amount. Amount should match the L2 supply.
+     *         Function signature complies by Bridged USDC Standard.
```
https://github.com/OffchainLabs/token-bridge-contracts/pull/87/commits/25672c7077deaa2eae1cbc94d95a0e48d9dd4362
